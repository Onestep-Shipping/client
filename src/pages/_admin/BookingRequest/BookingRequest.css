.booking-confirmation-container {
  margin-top: 30px;
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  margin-bottom: 30px;
}

.confirmation-info-container {
  width: 60%;
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 30px;
}

.booking-instruction-detail {
  height: 100%;
  width: 75%;
  background-color: rgb(255, 255, 255);
  border-left: 1px solid rgb(207, 207, 207);
  overflow: hidden; 
  overflow-y: scroll;
}

.booking-confirmation-input {
  width: 200px;
  font-size: 16px;
  padding: 8px 15px;
  border-radius: 4px;
  border: none;
  outline: none;
  border: solid #474747 1px;
  background-color: white;
  box-shadow: none;
}

.commodity-input {
  width: 70%;
  font-size: 16px;
  padding: 8px 15px;
  border-radius: 4px;
  border: none;
  outline: none;
  border: solid #474747 1px;
  background-color: white;
  box-shadow: none;
}

.quantity-input {
  width: 20%;
  font-size: 16px;
  padding: 8px 15px;
  border-radius: 4px;
  border: none;
  outline: none;
  border: solid #474747 1px;
  background-color: white;
  box-shadow: none;
}

.quantity-and-container {
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
}

.quantity-and-container-row {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: space-between;
  margin-bottom: 20px;
}