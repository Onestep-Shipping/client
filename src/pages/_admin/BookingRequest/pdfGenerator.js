import jsPDF from 'jspdf';
import moment from 'moment';
import Utils from '../../../utils/Helpers.js';

const doc = new jsPDF("p", "pt", 'letter');

const centeredText = (text, y) => {
    let textWidth = doc.getStringUnitWidth(text) * doc.internal.getFontSize() / doc.internal.scaleFactor;
    let textOffset = (doc.internal.pageSize.width - textWidth) / 2;
    doc.text(textOffset, y, text);
}

const preview = info => {
  doc.setProperties({
    title: 'Booking Confirmation #' + info.confirmation.bookingNo,
    subject: 'Info about PDF',
    author: 'Scarlet Nguyen',
    keywords: 'generated, javascript, web 2.0, ajax',
    creator: 'OneStep Shipping'
  });

  doc.setFont('helvatica');
  doc.setFontSize(16);
  doc.setFontType('bolditalic');
  doc.text(430, 60, "OneStep Shipping");

  doc.setFont('courier');
  doc.setFontSize(20);
  doc.setFontType('bold');
  centeredText(
    Utils.ordinalSuffixOf(info.confirmation.timeReceived) + 
    ' Booking Confirmation', 
    100
  );

  doc.setFontSize(12);
  doc.setFontType('normal');
  doc.text(50, 140, "TO: " + info.company.name);
  doc.text(50, 158, "    " + info.company.address.street);
  doc.text(50, 173, "    " + info.company.address.city);
  doc.text(50, 188, "    " + info.company.address.country);

  doc.setFontSize(10);
  doc.text(310, 130, "Our Reference No.:                1234567");
  doc.text(310, 145, "Contact          :              Rose Phan");
  doc.text(310, 160, "Tel              :      +1 (604) 500-4934");
  doc.text(310, 175, "Email            :   rosephan99@gmail.com");
  doc.text(310, 190, "Issue Date       :             " + moment().format("L"));

  doc.line(50, 200, 560, 200);

  doc.setFontSize(11);
  doc.text(80, 230, "From: " + info.schedule.route.startLocation);
  doc.text(350, 230, "To: " + info.schedule.route.endLocation);

  doc.text(80, 250, "ETD: " + info.confirmation.etd);
  doc.text(350, 250, "ETA: " + info.confirmation.eta);

  doc.text(80, 290, "Commodity: " + info.booking.commodity);

  doc.text(350, 290, "Q'ty");
  doc.text(420, 290, "Container Type");

  info.booking.containers.map((row, i) => function() {
    doc.text(350, 307 + (i * 17), "" + row.quantity);
    doc.text(420, 307 + (i * 17), row.containerType);
  })

  doc.setFontSize(12);
  doc.setFontType('bold');
  doc.text(80, 330, "Booking No.: " + info.confirmation.bookingNo);

  doc.setFontSize(11);
  doc.setFontType('normal');
  doc.text(80, 360, "Carrier: " + info.schedule.route.carrier);

  doc.text(80, 400, "Terminal Cut-off: " + info.confirmation.terminaCutoff);
  doc.text(80, 420, "Document Cut-off: " + info.confirmation.docCutoff);
  doc.text(80, 440, "VGM Cut-off     : " + info.confirmation.vgmCutoff);

  doc.setFontSize(12);
  doc.setFontType('bold');
  doc.text(80, 483, "Empty Pickup Location");
  doc.text(350, 483, "Return Location");

  doc.setFontType('normal');
  doc.setFontSize(10);
  doc.text(80, 500, info.confirmation.pickUpLocation.street);
  doc.text(350, 500, info.confirmation.returnLocation.street);

  doc.text(80, 515, info.confirmation.pickUpLocation.city);
  doc.text(350, 515, info.confirmation.returnLocation.city);

  doc.text(80, 530, info.confirmation.pickUpLocation.country);
  doc.text(350, 530, info.confirmation.returnLocation.country);

  doc.line(50, 560, 560, 560);

  doc.setFontSize(12);
  doc.setFontType('bold');
  doc.text(50, 580, 'Note:');
  doc.setFontSize(11);
  doc.setFontType('normal');
  doc.text(50, 605, 'Terminal cut-off may change. The trucker has to check with the terminal prior');
  doc.text(50, 625, 'to return.');

  doc.text(50, 670, 'Best regards,');
  doc.text(50, 710, 'OneStep Shipping.');

  window.open(doc.output('bloburl'), '_blank');
}

const uploadToServer = () => {
  return doc.output('blob');
}

export default { preview, uploadToServer };