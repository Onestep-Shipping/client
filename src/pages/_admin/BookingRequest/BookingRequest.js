import React, {useState} from 'react';
import './BookingRequest.css';

import Header from '../../../components/Header/Header.js';
import BookingDisplay from '../../../components/BookingDisplay/BookingDisplay.js';
import UserList from '../../../components/UserList/UserList.js';
import styles from '../../../components/ScheduleForm/ScheduleFormMin.module.css';
import DATA from '../../../data/ScheduleDetailsData.js';
import { InfoRow, ContainerDetail, Address, MiniDatePicker, MiniDatePickerTime } from '../Helpers.js';
import PdfGenerator from './PdfGenerator.js';
import FileUploadService from '../../../services/FileUploadService.js';
import GET_BOOKING_REQUEST from '../../../apollo/queries/GetBookingRequestQuery.js';
import { useQuery } from '@apollo/react-hooks';
import moment from 'moment';

const formatISOString = iso => {
  return moment(iso).utc().format('MM/DD/YYYY');
}

const BookingRequest = () => {
  const [currentBookingIndex, setCurrentBookingIndex] = useState(0);

  const { loading, error, data } = useQuery(GET_BOOKING_REQUEST, {
    fetchPolicy: 'cache-and-network'
  });

  const today = new Date();
  const myToday = new Date(today.getFullYear(), today.getMonth(), today.getDate(), 0, 0, 0);
  const [terminalDate, setTerminalDate] = useState(myToday);
  const [docDate, setDocDate] = useState(myToday);
  const [vgmDate, setVgmDate] = useState(myToday);
  const [etdDate, setEtdDate] = useState(DATA[currentBookingIndex].startDate);
  const [etaDate, setEtaDate] = useState(DATA[currentBookingIndex].endDate);

  const handleIndexChange = newInd => {
    setCurrentBookingIndex(newInd);
    setTerminalDate(myToday);
    setDocDate(myToday);
    setVgmDate(myToday);
    [...document.getElementsByTagName("input")].forEach(node => node.value = "");
  }

  const handlePreview = e => {
    e.preventDefault();
    PdfGenerator.preview(createInfoObject(e));
  }

  const handleUpload = e => {
    e.preventDefault();
    const blob = PdfGenerator.uploadToServer();
    const pdf = new File(
      [blob], 
      'Booking Confirmation #' + document.getElementById("bookingNo").value + '.pdf', 
      { type: 'application/pdf' } 
    )
    const formData = new FormData();
    formData.append("file", pdf);

    FileUploadService.uploadFile(formData)
      .then(res => {
        alert(res.data.fileLocation);
      })
      .catch(e => {
        console.log(e.response);
      });
  }

  if (loading) return <p>Loading...</p>;
  if (error) return <p>Error :(</p>;

  const shipments = data.getAllShipments;

  const  { bookingRequest, schedule, bookedBy } = shipments[currentBookingIndex];

  const quote = schedule.route.quoteHistory
    .filter(quote => Date.parse(quote.validity.startDate) <=  Date.parse(schedule.startDate))
    .slice(0, 1)[0];

  const createInfoObject = e => {
    const { bookingNo, 
            etd, eta, terminal, doc, vgm,
            pickupLocationStreet, pickupLocationCity, pickupLocationCountry,
            returnLocationStreet, returnLocationCity, returnLocationCountry 
          } = e.target;
    
    return {
      company: bookedBy,
      schedule: DATA[currentBookingIndex],
      booking: bookingRequest,
      bookingNo: bookingNo.value,
      etd: etd.value,
      eta: eta.value,
      terminalDate: terminal.value, 
      docDate: doc.value, 
      vgmDate: vgm.value,
      pickupLine1: pickupLocationStreet.value,
      pickupLine2: pickupLocationCity.value,
      pickupLine3: pickupLocationCountry.value,
      returnLine1: returnLocationStreet.value,
      returnLine2: returnLocationCity.value,
      returnLine3: returnLocationCountry.value,
    }
  }

  return (
    <div className="homepage-container">
      <Header />
      <div className="bol-instruction-container">
        <UserList  
          setInd={handleIndexChange}
          opt={shipments} type="booking"
        />
        <div className="booking-instruction-detail"> 
          <div className="booking-id-container">
            <h1>Booking Request</h1>
          </div>
          <div className="customer-info-container">
            <text>Contact: {bookedBy.personInCharge.name}</text>
            <text>Email: {bookedBy.email}</text>
            <text>
              {formatISOString(bookingRequest.form.createdAt)}
            </text>
          </div>

          <div className="form-container">
            <h2>Schedule</h2>
            <BookingDisplay  schedule={schedule} quote={quote} fields={8} />
          </div>

          <div className="form-container">
            <div className="booking-details-container">
              <InfoRow label="Commodity" value={bookingRequest.form.commodity} />
              <InfoRow label="HS Code" value={bookingRequest.form.hsCode} />
              <InfoRow label="Shipment Detail" value="" />
              <div className="shipment-detail-row">
                {bookingRequest.form.containers.map((row, ind) => (
                 <ContainerDetail key={ind} container={row} ind={ind} />
                ))}
              </div>
              <InfoRow label="Payment Term" value={bookingRequest.form.paymentTerm} />
              <InfoRow 
                label="CAED/AES filling by OneStep" 
                value={bookingRequest.form.autoFilling ? "Yes" : "No"} 
              />
            </div>
          </div>

          {!bookingRequest.status === "Received" &&
          <form className="booking-confirmation-container" onSubmit={handlePreview}>
            <h2>Booking Confirmation</h2>
            <div className="confirmation-info-container">
              <text className={styles.scheduleLabel}>Booking No.</text>
              <input
                type="text"
                id="bookingNo"
                name="bookingNo"
                className="commodity-input"
                placeholder="i.e. 1234567"
                required />
            </div>
            <div className="confirmation-info-container">
              <text className={styles.scheduleLabel}>ETD</text>
              <text className={styles.scheduleLabel}>ETA</text>
            </div>
            <div className="confirmation-info-container">
              <MiniDatePicker name="etd" value={new Date(etdDate)} action={setEtdDate} />
              <MiniDatePicker name="eta" value={new Date(etaDate)} action={setEtaDate} id="align-right"/>
            </div>
            <div className="confirmation-info-container">
              <text className={styles.scheduleLabel}>Terminal Cut-off</text>
              <text className={styles.scheduleLabel}>Document Cut-off</text>
              <text className={styles.scheduleLabel}>VGM Cut-off</text>
            </div>
            <div className="confirmation-info-container">
              <MiniDatePickerTime name="terminal" value={terminalDate} action={setTerminalDate} />
              <MiniDatePickerTime name="doc" value={docDate} action={setDocDate} />
              <MiniDatePickerTime name="vgm" value={vgmDate} action={setVgmDate} />
            </div>

            <div className="confirmation-info-container">
              <text className={styles.scheduleLabelSpecial}>Empty Pickup Location</text>
              <text className={styles.scheduleLabelSpecial}>Return Location</text>  
            </div>
            <Address type="Street" />
            <Address type="City" />
            <Address type="Country" />

            <div className="bol-button-form">
              <input id="left-button" type="submit" className="result-button" value="Generate PDF" />
              <button  className="result-button" onClick={handleUpload}>
                Send to Customer
              </button>
            </div>
          </form>}
        </div>
      </div>
    </div>
  );
};

export default BookingRequest;