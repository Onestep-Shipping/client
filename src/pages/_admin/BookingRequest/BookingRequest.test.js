import { ApolloProvider } from '@apollo/react-hooks';
import BookingRequest from './BookingRequest';
import React from 'react';
import { shallow } from 'enzyme';

let wrapper = shallow(
  <ApolloProvider>
    <BookingRequest />
  </ApolloProvider>
);

describe('BookingRequest', () => {
  it('should render the BookingRequest Component correctly', () => {   
    expect(BookingRequest).toMatchSnapshot();
  });
});