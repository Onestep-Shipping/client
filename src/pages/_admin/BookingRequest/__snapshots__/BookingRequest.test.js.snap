// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BookingRequest should render the BookingRequest Component correctly 1`] = `[Function]`;
